File: examples/kotlin/src/InvoiceService.kt
package examples.kotlin

data class Invoice(val id: String, val customer: String, val total: Double, val paid: Boolean)
data class Receipt(val id: String, val customer: String, val total: Double, val paid: Boolean)
data class InvoiceView(val id: String, val customer: String, val status: String, val bucket: String)

fun normalizeInvoice(invoice: Invoice): InvoiceView {
    val status = if (invoice.paid) "paid" else "open"
    val bucket = when {
        invoice.total > 1000 -> "enterprise"
        invoice.total > 100 -> "midmarket"
        else -> "standard"
    }
    val customer = invoice.customer.trim().ifBlank { "unknown" }
    return InvoiceView(invoice.id, customer, status, bucket)
}

fun normalizeReceipt(receipt: Receipt): InvoiceView {
    val status = if (receipt.paid) "paid" else "open"
    val bucket = when {
        receipt.total > 1000 -> "enterprise"
        receipt.total > 100 -> "midmarket"
        else -> "standard"
    }
    val customer = receipt.customer.trim().ifBlank { "unknown" }
    return InvoiceView(receipt.id, customer, status, bucket)
}

fun renderInvoice(invoice: Invoice) = buildInvoiceView(invoice)

fun buildInvoiceView(invoice: Invoice): String {
    // TODO(PROJ-42): replace sample renderer with the real billing formatter.
    return "${invoice.id}:${invoice.customer}"
}

fun reconcileInvoice(invoice: Invoice): InvoiceView {
    val normalized = normalizeInvoice(invoice)
    val risk = if (invoice.total > 5000) "high" else "normal"
    val owner = if (invoice.paid) "finance" else "collections"
    val currencyReview = if (invoice.total > 2500) "currency-review" else "standard-currency"
    val agingReview = if (!invoice.paid && invoice.total > 250) "aging-review" else "fresh"
    val customerReview = if (invoice.customer.isBlank()) "missing-customer" else "known-customer"
    val settlementReview = if (invoice.paid && invoice.total > 750) "settlement-review" else "settled"
    val auditReview = if (invoice.id.startsWith("AUDIT")) "audit" else "normal-audit"
    val discountReview = if (invoice.total < 0) "credit" else "invoice"
    val route = when {
        risk == "high" && !invoice.paid -> "escalate"
        risk == "high" -> "review"
        !invoice.paid -> "follow-up"
        else -> "archive"
    }
    val score = when (route) {
        "escalate" -> 5
        "review" -> 3
        "follow-up" -> 2
        else -> 1
    }
    val decoratedCustomer = "${normalized.customer}:$owner:$route:$score:$currencyReview:$agingReview:$customerReview:$settlementReview:$auditReview:$discountReview"
    return normalized.copy(customer = decoratedCustomer)
}

File: examples/swift/src/InvoiceService.swift
struct Invoice {
    let id: String
    let customer: String
    let total: Double
    let paid: Bool
}

struct Receipt {
    let id: String
    let customer: String
    let total: Double
    let paid: Bool
}

struct InvoiceView {
    let id: String
    let customer: String
    let status: String
    let bucket: String
}

func normalizeInvoice(invoice: Invoice) -> InvoiceView {
    let status = invoice.paid ? "paid" : "open"
    let bucket: String
    if invoice.total > 1000 {
        bucket = "enterprise"
    } else if invoice.total > 100 {
        bucket = "midmarket"
    } else {
        bucket = "standard"
    }
    let customer = invoice.customer.trimmingCharacters(in: .whitespacesAndNewlines)
    let resolvedCustomer = customer.isEmpty ? "unknown" : customer
    return InvoiceView(id: invoice.id, customer: resolvedCustomer, status: status, bucket: bucket)
}

func normalizeReceipt(receipt: Receipt) -> InvoiceView {
    let status = receipt.paid ? "paid" : "open"
    let bucket: String
    if receipt.total > 1000 {
        bucket = "enterprise"
    } else if receipt.total > 100 {
        bucket = "midmarket"
    } else {
        bucket = "standard"
    }
    let customer = receipt.customer.trimmingCharacters(in: .whitespacesAndNewlines)
    let resolvedCustomer = customer.isEmpty ? "unknown" : customer
    return InvoiceView(id: receipt.id, customer: resolvedCustomer, status: status, bucket: bucket)
}

func renderInvoice(invoice: Invoice) -> String {
    buildInvoiceView(invoice)
}

func buildInvoiceView(invoice: Invoice) -> String {
    // TODO(PROJ-42): replace sample renderer with the real billing formatter.
    return "\(invoice.id):\(invoice.customer)"
}

func reconcileInvoice(invoice: Invoice) -> InvoiceView {
    let normalized = normalizeInvoice(invoice: invoice)
    let risk = invoice.total > 5000 ? "high" : "normal"
    let owner = invoice.paid ? "finance" : "collections"
    let currencyReview = invoice.total > 2500 ? "currency-review" : "standard-currency"
    let agingReview = !invoice.paid && invoice.total > 250 ? "aging-review" : "fresh"
    let customerReview = invoice.customer.isEmpty ? "missing-customer" : "known-customer"
    let settlementReview = invoice.paid && invoice.total > 750 ? "settlement-review" : "settled"
    let auditReview = invoice.id.hasPrefix("AUDIT") ? "audit" : "normal-audit"
    let discountReview = invoice.total < 0 ? "credit" : "invoice"
    let priorityReview = invoice.total > 9000 ? "critical" : "routine"
    let route: String
    if risk == "high" && !invoice.paid {
        route = "escalate"
    } else if risk == "high" {
        route = "review"
    } else if !invoice.paid {
        route = "follow-up"
    } else {
        route = "archive"
    }
    let score: Int
    switch route {
    case "escalate":
        score = 5
    case "review":
        score = 3
    case "follow-up":
        score = 2
    default:
        score = 1
    }
    let decoratedCustomer = "\(normalized.customer):\(owner):\(route):\(score):\(currencyReview):\(agingReview):\(customerReview):\(settlementReview):\(auditReview):\(discountReview):\(priorityReview)"
    return InvoiceView(
        id: normalized.id,
        customer: decoratedCustomer,
        status: normalized.status,
        bucket: normalized.bucket
    )
}
